### PR TITLE
Beam search

### PR DESCRIPTION
**Pull request by lpworld**
Add beam search options. Current default model could cause disturbance between the padding words, though. @hitvoice

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -252,14 +253,28 @@ def generate(self, hidden, length, sample=False, temp=1.0):
         for i in range(torch.max(length.data)):
             output, state = self.decoder(inputs, state)
             overvocab = self.linear(output.squeeze(1))
+            sen_score = {}
 
             if not sample:
-                vals, indices = torch.max(overvocab, 1)
+                    vals, indices = torch.topk(overvocab, beam_size, 1)
```
Comment (hitvoice) [inline models.py]: extra white spaces should be cleaned

File: models.py [hunk 2]
```diff
@@ -252,14 +253,28 @@ def generate(self, hidden, length, sample=False, temp=1.0):
...
+            for index in range(beam_size):
+                string = []
+                string = all_indices.copy()
+                string.append(torch.index_select(indices, 1, Variable(torch.LongTensor([index]))))
+                string = torch.cat(string, 1)
+                string = string.data.cpu().numpy()
+                string = [[vocab[x] for x in idx] for idx in string]
+                sen_score[index] = sum(score(x) for x in string)
+
+            print(sen_score)
```
Comment (hitvoice) [inline models.py]: "print" should not be left in the code.

File: models.py [hunk 3]
```diff
@@ -252,14 +253,28 @@ def generate(self, hidden, length, sample=False, temp=1.0):
...
+                    vals, indices = torch.topk(overvocab, beam_size, 1)
             else:
                 # sampling
                 probs = F.softmax(overvocab / temp)
-                indices = torch.multinomial(probs, 1)
-
+                indices = torch.multinomial(probs, beam_size)
+
+            for index in range(beam_size):
+                string = []
```
Comment (hitvoice) [inline models.py]: This line seems unnecessary since `string` will be overwritten by a new value in the next line.

File: models.py [hunk 4]
```diff
@@ -252,14 +253,28 @@ def generate(self, hidden, length, sample=False, temp=1.0):
...
+                string = all_indices.copy()
+                string.append(torch.index_select(indices, 1, Variable(torch.LongTensor([index]))))
```
Comment (hitvoice) [inline models.py]: `torch.index_select(indices, 1, Variable(torch.LongTensor([index])))` is equal to `indices[:, index]`
